[export] Add custom op profiles and generate meta kernel

angelayi opened the pull request:
Example output:
```
1. Missing fake kernel.
    torch.ops.mylib.foo2.default is missing a fake kernel implementation.

    Please refer to https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit#heading=h.ahugy69p2jmz for more detailed instructions on how to implement a fake kernel.

    Here is an example of a fake kernel for your op, but it might not be correct for all use cases:
    ```
import torch.utils._pytree as pytree

@torch.library.register_fake("mylib::foo3")
def _(*args, **kwargs):
    ctx = torch.library.get_ctx()

    def match(tensor, dim, dtype, device, layout):
        # TODO: check strides
        return (
            tensor.dim() == dim
            and tensor.dtype == dtype
            and tensor.device == device
            and tensor.layout == layout
        )

    # Generate fake tensor for each output
    def get_fake_out(dim, sorted_strides, dtype, device, layout):
        fake_shape = [ctx.new_dynamic_size() for _ in range(dim)]

        fake_strides = [-1] * dim
        fake_stride = 1
        for idx in sorted_strides:
            fake_strides[idx] = fake_stride
            fake_stride = fake_stride * fake_shape[idx]

        return torch.empty_strided(
            size=fake_shape, stride=fake_strides, dtype=dtype, device=device, layout=layout
        )
    
    flat_args, args_spec = pytree.tree_flatten(args)
    
    if (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 3, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 3, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(3, [2, 1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    elif (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 2, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 2, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(2, [1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    else:
        raise NotImplementedError("Input type not seen before")
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #140048
* #139384

angelayi committed:
[export] Add custom op profiles and generate meta kernel

[ghstack-poisoned]

angelayi committed:
Update on "[export] Add custom op profiles and generate meta kernel"

Example output:
```
1. Missing fake kernel.
    torch.ops.mylib.foo2.default is missing a fake kernel implementation.

    Please refer to https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit#heading=h.ahugy69p2jmz for more detailed instructions on how to implement a fake kernel.

    Here is an example of a fake kernel for your op, but it might not be correct for all use cases:
    ```
    import torch.utils._pytree as pytree

    torch.library.register_fake("mylib::foo2")
    def mylib_foo2_default_fake(*args, **kwargs):
        ctx = torch.library.get_ctx()

        # Generate fake tensor for each output
        def get_fake_out(dim, sorted_strides, dtype, device, layout):
            fake_shape = [ctx.new_dynamic_size() for _ in range(dim)]

            fake_strides = [-1] * dim
            fake_stride = 1
            for idx in sorted_strides:
                fake_strides[idx] = fake_stride
                fake_stride = fake_stride * fake_shape[idx]

            return torch.empty_strided(
                size=fake_shape, stride=fake_strides, dtype=dtype, device=device, layout=layout
            )

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(2, [1, 0], torch.float32, "cpu", torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )
    ```
```




[ghstack-poisoned]

pianpwk commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -2788,4 +2786,29 @@ def unsupported(reason: str) -> None:
+        }
+
+    dtrace_structured(
+        "generated_fake_kernel",
```
nit: could we rename this case to "missing_fake_kernel" or something similar? https://github.com/pytorch/pytorch/pull/139766 will also generate fake kernels, but for the mismatched case

angelayi commented on torch/export/_draft_export.py:
```diff
@@ -296,6 +384,34 @@ def draft_export(
+            failures.append(
+                FailureReport(
+                    FailureType.MISSING_FAKE_KERNEL,
+                    {"op": op, "op_profiles": list(seen_profiles)},
```
I was thinking we can use the op_profiles during retracing? Or in cases where we don't have real tensor tracing.

zou3519 replied:
It's probably worth sitting down and ironing out the "ideal UX" for reuse of the custom op profile. Maybe the ideal UX is "we get the user to check in the meta to the codebase, so they do not need to serialize and pass the profile around".

angelayi commented on torch/export/_draft_export.py:
````diff
@@ -89,11 +91,63 @@ def __repr__(self) -> str:
+        return pytree.tree_unflatten(
+            flat_fake_out, pytree.treespec_loads(\'{op_profiles[0].out_spec}\')
+        )
+    ```
````
Yeah for now I only used the first op profile.. it seemed a little hacky to generate the ifs, but thinking more about it, it shouldn't be too hard..

angelayi replied:
updated!

angelayi committed:
Update on "[export] Add custom op profiles and generate meta kernel"

Example output:
```
1. Missing fake kernel.
    torch.ops.mylib.foo2.default is missing a fake kernel implementation.

    Please refer to https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit#heading=h.ahugy69p2jmz for more detailed instructions on how to implement a fake kernel.

    Here is an example of a fake kernel for your op, but it might not be correct for all use cases:
    ```
import torch.utils._pytree as pytree

torch.library.register_fake("mylib::foo3")
def _(*args, **kwargs):
    ctx = torch.library.get_ctx()

    def match(tensor, dim, dtype, device, layout):
        # TODO: check strides
        return (
            tensor.dim() == dim
            and tensor.dtype == dtype
            and tensor.device == device
            and tensor.layout == layout
        )

    # Generate fake tensor for each output
    def get_fake_out(dim, sorted_strides, dtype, device, layout):
        fake_shape = [ctx.new_dynamic_size() for _ in range(dim)]

        fake_strides = [-1] * dim
        fake_stride = 1
        for idx in sorted_strides:
            fake_strides[idx] = fake_stride
            fake_stride = fake_stride * fake_shape[idx]

        return torch.empty_strided(
            size=fake_shape, stride=fake_strides, dtype=dtype, device=device, layout=layout
        )
    
    flat_args, args_spec = pytree.tree_flatten(args)
    
    if (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 3, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 3, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(3, [2, 1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    elif (
        args_spec == pytree.treespec_loads('[1, {"type": "builtins.tuple", "context": "null", "children_spec": [{"type": "builtins.tuple", "context": "null", "children_spec": [{"type": null, "context": null, "children_spec": []}, {"type": null, "context": null, "children_spec": []}]}, {"type": "builtins.dict", "context": "[]", "children_spec": []}]}]') 
        and match(flat_args[0], 2, torch.float32, torch.device("cpu"), torch.strided)
        and match(flat_args[1], 2, torch.float32, torch.device("cpu"), torch.strided)
    ):

        # Loop through all the flat outputs from the profiling
        flat_fake_out = [
            get_fake_out(2, [1, 0], torch.float32, torch.device("cpu"), torch.strided),
        ]

        # Unflatten the outputs to be the same output structure as the eager model
        return pytree.tree_unflatten(
            flat_fake_out, pytree.treespec_loads('[1, {"type": null, "context": null, "children_spec": []}]')
        )

    else:
        raise NotImplementedError("Input type not seen before")
```




[ghstack-poisoned]

pianpwk commented on torch/export/_draft_export.py:
```diff
@@ -89,11 +98,105 @@ def __repr__(self) -> str:
+            profile_casing = ""
+
+            for i, op_profile in enumerate(op_profiles):
+                check_arg_string = '            and match(flat_args[{idx}], {dim}, {dtype}, torch.device("{device}"), {layout})'
```
nit: maybe add keywords? `size={dim}, dtype={dtype}, ...`, same for the bottom?

pianpwk commented on torch/export/_draft_export.py:
```diff
@@ -89,11 +98,105 @@ def __repr__(self) -> str:
+        {"if" if i == 0 else "elif"} (
+            args_spec == pytree.treespec_loads(\'{op_profile.args_spec}\')
+{check_args_str}
+        ):
```
Oh, is it the `_custom_ops_profile.py` that's still only in fbcode? But why is that not in OSS now? Sorry, I'm just confused.

angelayi replied:
the op profiles are created [further down in this PR](https://github.com/pytorch/pytorch/pull/140048/files#diff-86001d9642d74824f75bd856be741e67deb1baa8c4a07a43581f4ddd72e809f9R444-R460)
